### ISSUE TITLE
Inner workings of allreduce while computing AUC

### ISSUE THREAD

**Issue by patrickcho168** [uber/horovod#444]
Hi @alsrgv 

I'm trying to compute validation metrics and I'm wondering how allreduce combines the metrics calculated on different GPUs.

From https://github.com/uber/horovod/issues/362
```
auc, auc_op = tf.metrics.auc(labels=lbl, predictions=p, curve='ROC')
auc_avg = hvd.allreduce(auc)
```

In this case, are we averaging the auc values computed on different machines? Taking an average of auc values in such a manner will not be give an accurate overall auc value. 

Or are we averaging the TP, FP, TN, FN values for each threshold and recomputing the auc? Doing so would give an accurate overall auc value.

Also, if I want to compute the overall loss, can I perform the same operation? That is:
```
loss = tf.reduce_mean(losses)
loss_avg = hvd.allreduce(loss)
```

Thanks!

**Comment by patrickcho168** [comment]
Running the following experiment:
```python
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
import os
import tensorflow as tf
import horovod.tensorflow as hvd
import numpy as np
from sklearn.metrics import roc_auc_score
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'

# Expected Result: 0.70312494

# Horovod Stuff
hvd.init()
is_chief = hvd.rank() == 0
num_examples_per_gpu = 4
num_gpus = 8

# Generate Labels
high_labels = [0] + [1 for i in range(num_examples_per_gpu-1)]
low_labels  = [0 for i in range(num_examples_per_gpu-1)] + [1]
all_labels  = []
for _ in range(num_examples_per_gpu):
    all_labels += high_labels
for _ in range(num_examples_per_gpu):
    all_labels += low_labels
if is_chief:
    print("labels:", all_labels)

# Generate Probs
total_examples = num_examples_per_gpu * num_gpus
all_probs = [(total_examples-i)/(total_examples+1) for i in range(total_examples)]
if is_chief:
    print("probs:", all_probs)
all_labels = np.array(all_labels)
all_probs = np.array(all_probs)

if is_chief:
    print("Expected SKLearn Result")
    print(roc_auc_score(all_labels, all_probs))

local_size = hvd.local_size()
local_rank = hvd.local_rank()

one_labels = all_labels[local_rank*num_examples_per_gpu:(local_rank+1)*num_examples_per_gpu]
one_probs = all_probs[local_rank*num_examples_per_gpu:(local_rank+1)*num_examples_per_gpu]

one_labels_placeholder = tf.placeholder(tf.float64, shape=(None,))
one_probs_placeholder = tf.placeholder(tf.float64, shape=(None,))

hvd_auc, hvd_auc_op = tf.metrics.auc(labels=one_labels_placeholder, predictions=one_probs_placeholder, curve='ROC')
hvd_summary_op = tf.summary.scalar("AUC", hvd_auc_op)
hvd_auc_avg = hvd.allreduce(hvd_auc)

### FIX HERE
auc_vars = tf.get_collection(tf.GraphKeys.LOCAL_VARIABLES)
print_auc_vars = tf.Print(auc_vars, auc_vars)
auc_vars_avg = [tf.assign(v, hvd.allreduce(v)) for v in auc_vars]
### FIX END

config = tf.ConfigProto(allow_soft_placement=True) #, log_device_placement=True
config.gpu_options.allow_growth = True
config.gpu_options.visible_device_list = str(hvd.local_rank())

def run_hvd():
    feed_dict = {
        one_labels_placeholder: one_labels,
        one_probs_placeholder: one_probs,
    }

    with tf.Session(config=config) as sess:
        sess.run(tf.local_variables_initializer())
        sess.run([hvd_auc_op], feed_dict=feed_dict)
        print(local_rank, sess.run([hvd_auc_avg, hvd_auc], feed_dict=feed_dict))

def run_real():
    feed_dict = {
        one_labels_placeholder: all_labels,
        one_probs_placeholder: all_probs,
    }

    with tf.Session(config=config) as sess:
        sess.run(tf.local_variables_initializer())
        sess.run([hvd_auc_op], feed_dict=feed_dict)
        print(local_rank, sess.run([hvd_auc_avg, hvd_auc], feed_dict=feed_dict))

def run_hvd_fixed():
    feed_dict = {
        one_labels_placeholder: one_labels,
        one_probs_placeholder: one_probs,
    }

    with tf.Session(config=config) as sess:
        sess.run(tf.local_variables_initializer())
        sess.run([hvd_auc_op], feed_dict=feed_dict)
        sess.run(auc_vars_avg)
        print(local_rank, sess.run([hvd_auc_avg, hvd_auc], feed_dict=feed_dict))

if is_chief:
    print("Current Horovod Result")
run_hvd()
if is_chief:
    print("Expected Horovod Result")
run_real()
if is_chief:
    print("Fixed Horovod Result")
run_hvd_fixed()
```

Output:
```
Expected SKLearn Result
0.703125
Current Horovod Result
1 [6.666658e-07, 3.333329e-07]
2 [6.666658e-07, 3.333329e-07]
3 [6.666658e-07, 3.333329e-07]
4 [6.666658e-07, 9.999987e-07]
5 [6.666658e-07, 9.999987e-07]
6 [6.666658e-07, 9.999987e-07]
7 [6.666658e-07, 9.999987e-07]
0 [6.666658e-07, 3.333329e-07]
Expected Horovod Result
7 [0.70312494, 0.70312494]
1 [0.70312494, 0.70312494]
2 [0.70312494, 0.70312494]
3 [0.70312494, 0.70312494]
4 [0.70312494, 0.70312494]
5 [0.70312494, 0.70312494]
6 [0.70312494, 0.70312494]
0 [0.70312494, 0.70312494]
Fixed Horovod Result
6 [0.7031248, 0.7031248]
7 [0.7031248, 0.7031248]
1 [0.7031248, 0.7031248]
2 [0.7031248, 0.7031248]
3 [0.7031248, 0.7031248]
4 [0.7031248, 0.7031248]
5 [0.7031248, 0.7031248]
0 [0.7031248, 0.7031248]
```

From these experiments, it seems like the correct way to do it is to do:
```
auc_vars = tf.get_collection(tf.GraphKeys.LOCAL_VARIABLES)
auc_vars_avg = [tf.assign(v, hvd.allreduce(v)) for v in auc_vars]
```
instead of 
```
hvd_auc_avg = hvd.allreduce(hvd_auc)
```

However, in actual practice, it might be fine to just average the auc scores, as long as the datasets are large and randomly shuffled.
